File: 2016/01/Main.kt
import java.io.File
import java.io.InputStream

import tcollier.AocExecutor

data class Complex(val real: Int = 0, val imag: Int = 0)

operator fun Complex.plus(c: Complex): Complex{
  return Complex(this.real + c.real, this.imag + c.imag)
}

operator fun Complex.times(c: Complex): Complex{
  return Complex(this.real * c.real - this.imag * c.imag, this.imag * c.real + this.real * c.imag)
}

fun part1Answer(input: List<String>): String {
  var position = Complex(0, 0)
  var bearing = Complex(0, 1)

  for (dir in input) {
    if (dir[0] == 'R') {
      bearing *= Complex(0, -1)
    } else if (dir[0] == 'L') {
      bearing *= Complex(0, 1)
    }

    val mag = dir.substring(1, dir.length).toInt()
    position += bearing * Complex(mag, 0)
  }
  return (Math.abs(position.real) + Math.abs(position.imag)).toString()
}

fun intersection(h1: Complex, h2: Complex, v1: Complex, v2: Complex): Complex? {
  var minX = h1.real
  var maxX = h2.real
  if (minX > maxX) {
    minX = minX xor maxX
    maxX = minX xor maxX
    minX = minX xor maxX
  }
  if (!(v1.real in minX..maxX)) {
    return null
  }
  var minY = v1.imag
  var maxY = v2.imag
  if (minY > maxY) {
    minY = minY xor maxY
    maxY = minY xor maxY
    minY = minY xor maxY
  }
  if (!(h1.imag in minY..maxY)) {
    return null
  }
  return Complex(v1.real, h1.imag)
}

fun part2Answer(input: List<String>): String {
  var position = Complex(0, 0)
  var bearing = Complex(0, 1)
  var visited = mutableSetOf<Complex>()
  visited.add(position)

  for (dir in input) {
    if (dir[0] == 'R') {
      bearing *= Complex(0, -1)
    } else if (dir[0] == 'L') {
      bearing *= Complex(0, 1)
    }

    val mag = dir.substring(1, dir.length).toInt()
    for (i in 1..mag) {
      position += bearing * Complex(1, 0)
      if (position in visited) {
        return (Math.abs(position.real) + Math.abs(position.imag)).toString()
      } else {
        visited.add(position)
      }
    }
  }
  return "Not Found"
}

fun main(args: Array<String>) {
    var lines = mutableListOf<String>()
    val inputStream: InputStream = File("2016/01/input.txt").inputStream()
    inputStream.bufferedReader().forEachLine { lines.add(it) }
    val input = lines[0].split(", ")
    val executor = AocExecutor(input, ::part1Answer, ::part2Answer)
    executor.run(args)
}

File: 2016/01/main.rs
use std::collections::HashSet;
use std::fs;
use std::env;
use std::ops;

mod util;

#[derive(Copy, Clone, Eq, Hash)]
struct Complex {
    real: i32,
    imag: i32,
}

impl ToString for Complex {
    fn to_string(&self) -> String {
        if self.imag == 0 {
            return self.real.to_string();
        } else if self.real == 0 {
            return format!("{}i", self.imag);
        } else {
            let sign = if self.imag < 0 { "-" } else { "+" };
            return format!("{} {} {}i", self.real, sign, self.imag.abs());
        }
    }
}

impl PartialEq for Complex {
    fn eq(&self, rhs: &Complex) -> bool {
        return self.real == rhs.real && self.imag == rhs.imag;
    }
}

impl ops::Add<Complex> for Complex {
    type Output = Complex;

    fn add(self, rhs: Complex) -> Complex {
        return Complex { real: self.real + rhs.real, imag: self.imag + rhs.imag };
    }
}

impl ops::Mul<i32> for Complex {
    type Output = Complex;

    fn mul(self, rhs: i32) -> Complex {
        return Complex { real: self.real * rhs, imag: self.imag * rhs };
    }
}

impl ops::Mul<Complex> for Complex {
    type Output = Complex;

    fn mul(self, rhs: Complex) -> Complex {
        return Complex {
            real: self.real * rhs.real - self.imag * rhs.imag,
            imag: self.imag * rhs.real + self.real * rhs.imag
        };
    }
}

struct Day1Solution {
    instructions: Vec<String>
}

impl util::Solution for Day1Solution {
    fn part1_result(&self) -> String {
        let mut position = Complex { real: 0, imag: 0 };
        let mut bearing = Complex { real: 0, imag: 1 };
        for instruction in &self.instructions {
            let direction = &instruction[0..1];
            let magnitude = instruction[1..].parse::<i32>().unwrap();
            let imag = if direction == "R" { -1 } else { 1 };
            bearing = bearing * Complex { real: 0, imag: imag };
            position = position + bearing * magnitude;
        }

        return (position.real.abs() + position.imag.abs()).to_string();
    }

    fn part2_result(&self) -> String {
        let mut position = Complex { real: 0, imag: 0 };
        let mut bearing = Complex { real: 0, imag: 1 };
        let mut visited: HashSet<Complex> = HashSet::new();
        visited.insert(position);
        for instruction in &self.instructions {
            let direction = &instruction[0..1];
            let magnitude = instruction[1..].parse::<i32>().unwrap();
            let imag = if direction == "R" { -1 } else { 1 };
            bearing = bearing * Complex { real: 0, imag: imag };
            for i in 0..magnitude {
                position = position + bearing;
                if visited.contains(&position) {
                    return (position.real.abs() + position.imag.abs()).to_string();
                } else {
                    visited.insert(position);
                }
            }
        }
        panic!("Not found")
    }
}

fn main() {
    let mut instructions: Vec<String> = Vec::new();
    let contents = fs::read_to_string("./2016/01/input.txt").expect("File not found");
    for line in contents.lines() {
        for inst in line.split(", ") {
            instructions.push(inst.to_string());
        }
    }

    let solution = Day1Solution { instructions: instructions };
    let executor = util::Executor::new(&solution as &util::Solution);
    let args = env::args().collect();
    executor.run(args);
}
